airbyte-integrations/connectors/source-apple-search-ads/source_apple_search_ads/apple_search_ads.yaml

```diff
@@ -0,0 +1,228 @@
+    http_method: "GET"
+    authenticator:
+      type: "OAuthAuthenticator"
+      token_refresh_endpoint: "https://appleid.apple.com/auth/oauth2/token?grant_type=client_credentials&scope=searchadsorg"
+      client_id: "{{ config['client_id'] }}"
+      client_secret: "{{ config['client_secret'] }}"
+      refresh_token: ""
+    request_options_provider:
+      request_headers:
+        X-AP-Context: "orgId={{ config['org_id'] }}"
+  retriever:
+    requester:
+      $ref: "*ref(definitions.requester)"
+      error_handler:
+        response_filters:
+          - http_codes: [ 500 ]
+            action: RETRY
+
+  base_stream:
+    selector:
+      extractor:
+        field_pointer: [ "data" ]
+    retriever:
+      $ref: "*ref(definitions.retriever)"
+      record_selector:
+        $ref: "*ref(definitions.base_stream.selector)"
+      paginator:
+        type: DefaultPaginator
+        pagination_strategy:
+          type: OffsetIncrement
+          page_size: 1000
+        page_size_option:
+          inject_into: request_parameter
+          field_name: limit
+        page_token_option:
+          field_name: offset
+          inject_into: request_parameter
+        url_base: "*ref(definitions.requester.url_base)"
+
+  campaigns_stream:
+    $ref: "*ref(definitions.base_stream)"
+    $options:
+      name: "campaigns"
+      primary_key: "id"
+      path: "/campaigns"
+
+  stream_slicer_campaign:
+    type: SubstreamSlicer
+    parent_stream_configs:
+      - stream: "*ref(definitions.campaigns_stream)"
+        parent_key: "id"
+        stream_slice_field: "campaign_id"
+
+  adgroups_stream:
+    $ref: "*ref(definitions.base_stream)"
+    $options:
+      name: "adgroups"
+      primary_key: "id"
+    retriever:
+      $ref: "*ref(definitions.base_stream.retriever)"
+      requester:
+        $ref: "*ref(definitions.requester)"
+        path: "/campaigns/{{ stream_slice.campaign_id }}/adgroups"
+      stream_slicer:
+        type: CartesianProductStreamSlicer
+        stream_slicers:
+          - "*ref(definitions.stream_slicer_campaign)"
+
+  stream_slicer_adgroup:
+    type: SubstreamSlicer
+    parent_stream_configs:
+      - stream: "*ref(definitions.adgroups_stream)"
+        parent_key: "id"
+        stream_slice_field: "adgroup_id"
+
+  keywords_stream:
+    $ref: "*ref(definitions.base_stream)"
+    $options:
+      name: "keywords"
+      primary_key: "id"
+    retriever:
+      $ref: "*ref(definitions.base_stream.retriever)"
+      requester:
+        $ref: "*ref(definitions.requester)"
+        path: "/campaigns/{{ stream_slice.parent_slice.campaign_id }}/adgroups/{{ stream_slice.adgroup_id }}/targetingkeywords"
+      stream_slicer:
+        type: CartesianProductStreamSlicer
+        stream_slicers:
+          - "*ref(definitions.stream_slicer_adgroup)"
+
+  stream_slicer_date:
+    type: "DatetimeStreamSlicer"
+    start_datetime:
+      datetime: "{{ config['start_date'] }}"
```

Comment:
I would like to do something like `max(config['start_date'], day_delta(-90))` as the daily reports are limited to 90 days in the past and trying to read beyond will trigger an error. 

But the `day_delta` [macro](https://github.com/airbytehq/airbyte/blob/master/airbyte-cdk/python/airbyte_cdk/sources/declarative/interpolation/macros.py#L86) is a datetime while I need a date. Am I allowed to add a macro? I don't want to clutter the macros with too specific cases.

Reply:
You can use datetime and use formatting to date, what do you think?

Reply:
Yes I've tried that a few weeks ago but could not figure how to put it in the interpolated string 😅  

However I just discovered this is not needed because there's actually no limitation in the `start_date`: setting it >=90 days does not throw any error (and returns data) so I don't need to limit this date parameter 👍